### ISSUE TITLE
Installation problems with deb and cmake

### ISSUE THREAD

**Issue by samuelbertolini** [mlpack/mlpack#3090]
Hi all! I am new here, and I am having troubles to install mlpack. I tried two ways: by "apt install mlpack-bin" and by cmake. Both is not working. Can anyone please help? Sorry for such a simple question.
I have installed all the dependencies, e.g. libarmadillo10, libboost, etc.

Errors output (apt install):

> sudo apt install mlpack-bin 
> Reading package lists... Done
> Building dependency tree... Done
> Reading state information... Done
> Some packages could not be installed. This may mean that you have
> requested an impossible situation or if you are using the unstable
> distribution that some required packages have not yet been created
> or been moved out of Incoming.
> The following information may help to resolve the situation:
> 
> The following packages have unmet dependencies:
>  mlpack-bin : Depends: libarmadillo9 (>= 1:9.900.1+dfsg) but it is not installable
>               Depends: libboost-program-options1.71.0 but it is not installable
>               Depends: libboost-serialization1.71.0 but it is not installable
>               Depends: libboost-test1.71.0 but it is not installable
>               Depends: libmlpack3 (>= 3.3.2-5) but it is not going to be installed
> E: Unable to correct problems, you have held broken packages.

Errors output (cmake):

> [ 45%] Built target mlpack_softmax_regression
> [ 45%] Built target mlpack_sparse_coding
> [ 45%] Building CXX object src/mlpack/tests/CMakeFiles/mlpack_catch_test.dir/main.cpp.o
> In file included from /usr/include/signal.h:328,
>                  from /home/samuel/Downloads/mlpack-3.4.2/src/mlpack/tests/catch.hpp:5354,
>                  from /home/samuel/Downloads/mlpack-3.4.2/src/mlpack/tests/main.cpp:16:
> /home/samuel/Downloads/mlpack-3.4.2/src/mlpack/tests/catch.hpp:7947:58: error: call to non-‘constexpr’ function ‘long int sysconf(int)’
>  7947 |     constexpr static std::size_t sigStackSize = 32768 >= MINSIGSTKSZ ? 32768 : MINSIGSTKSZ;
>       |                                                          ^~~~~~~~~~~
> In file included from /usr/include/boost/config/stdlib/libstdcpp3.hpp:78,
>                  from /usr/include/boost/config.hpp:48,
>                  from /usr/include/boost/serialization/serialization.hpp:13,
>                  from /home/samuel/Downloads/mlpack-3.4.2/src/mlpack/prereqs.hpp:86,
>                  from /home/samuel/Downloads/mlpack-3.4.2/src/mlpack/core.hpp:67,
>                  from /home/samuel/Downloads/mlpack-3.4.2/build/src/mlpack/tests/cotire/mlpack_catch_test_CXX_prefix.cxx:4,
>                  from /home/samuel/Downloads/mlpack-3.4.2/build/src/mlpack/tests/cotire/mlpack_catch_test_CXX_prefix.hxx:4:
> /usr/include/unistd.h:640:17: note: ‘long int sysconf(int)’ declared here
>   640 | extern long int sysconf (int __name) __THROW;
>       |                 ^~~~~~~
> In file included from /home/samuel/Downloads/mlpack-3.4.2/src/mlpack/tests/main.cpp:16:
> /home/samuel/Downloads/mlpack-3.4.2/src/mlpack/tests/catch.hpp:8006:45: error: size of array ‘altStackMem’ is not an integral constant-expression
>  8006 |     char FatalConditionHandler::altStackMem[sigStackSize] = {};
>       |                                             ^~~~~~~~~~~~
> make[2]: *** [src/mlpack/tests/CMakeFiles/mlpack_catch_test.dir/build.make:825: src/mlpack/tests/CMakeFiles/mlpack_catch_test.dir/main.cpp.o] Error 1
> make[1]: *** [CMakeFiles/Makefile2:4036: src/mlpack/tests/CMakeFiles/mlpack_catch_test.dir/all] Error 2
> make: *** [Makefile:160: all] Error 2

**Comment by eddelbuettel** [comment]
Which distribution and version are you using?  Which repo are you pointing at with `i.e.` can you show us what `apt-cache policy mlpack-bin` shows?

**Comment by eddelbuettel** [comment]
On an 'empty' Docker container (from R work) that runs Debian testing I have no issues:

```txt
root@35eeacb074ca:/work# apt update    # ignore the links to eddelbuettel.github.io/ppaR400
Get:1 http://deb.debian.org/debian testing InRelease [128 kB]
Ign:2 https://eddelbuettel.github.io/ppaR400 ./ InRelease                                                                                
Get:4 https://eddelbuettel.github.io/ppaR400 ./ Release [1,204 B]     
Ign:5 https://eddelbuettel.github.io/ppaR400 ./ Release.gpg                                                         
Get:6 http://deb.debian.org/debian experimental InRelease [75.4 kB] 
Get:7 https://eddelbuettel.github.io/ppaR400 ./ Packages [26.4 kB]         
Get:8 http://deb.debian.org/debian testing/main amd64 Packages [8,269 kB]
Get:3 http://cdn-fastly.deb.debian.org/debian sid InRelease [165 kB]            
Get:9 http://cdn-fastly.deb.debian.org/debian sid/main amd64 Packages [8,830 kB]
Get:10 http://deb.debian.org/debian experimental/main amd64 Packages [407 kB]      
Fetched 17.9 MB in 3s (6,169 kB/s)                                                       
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
140 packages can be upgraded. Run 'apt list --upgradable' to see them.
root@35eeacb074ca:/work# apt install mlpack-bin
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libarmadillo10 libarpack2 libboost-serialization1.74.0 libboost-test1.74.0 libmlpack3 libsuperlu5
The following NEW packages will be installed:
  libarmadillo10 libarpack2 libboost-serialization1.74.0 libboost-test1.74.0 libmlpack3 libsuperlu5 mlpack-bin
0 upgraded, 7 newly installed, 0 to remove and 140 not upgraded.
Need to get 27.7 MB of archives.
After this operation, 284 MB of additional disk space will be used.
Do you want to continue? [Y/n] 
```

**Comment by samuelbertolini** [comment]
I am trying in Ubuntu Impish

the commands below gives
```
~$ apt-cache policy mlpack-bin
mlpack-bin:
  Installed: (none)
  Candidate: 3.3.2-5
  Version table:
     3.3.2-5 500
        500 http://archive.ubuntu.com/ubuntu impish-proposed/universe amd64 Packages
```
and

**Comment by eddelbuettel** [comment]
I am too on another laptop but

```sh
edd@brad:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 21.10
Release:        21.10
Codename:       impish
edd@brad:~$ apt-cache policy mlpack-bin
N: Unable to locate package mlpack-bin
edd@brad:~$ 
```

**Comment by samuelbertolini** [comment]
Thanks in any case!

**Comment by conradsnicta** [comment]
> ```/home/samuel/Downloads/mlpack-3.4.2/src/mlpack/tests/catch.hpp:7947:58: error: call to non-‘constexpr’ function ‘long int sysconf(int)’```
> ```7947 | constexpr static std::size_t sigStackSize = 32768 >= MINSIGSTKSZ ? 32768 : MINSIGSTKSZ;```

This is due to a bug in an old version of catch.hpp from the Catch2 testing framework: https://github.com/catchorg/Catch2/issues/2178

mlpack 3.4.2 is pretty old by now.  Suggest using the git version of mlpack until the next official release.

@rcurtin @zoq @shrit Perhaps a quick interim release is warranted?  Say mlpack 3.4.3, where only the catch.hpp header is updated and the tests are disabled (they're not useful to users, as per https://github.com/mlpack/mlpack/pull/2926); no other changes, as that would take too long.

**Comment by samuelbertolini** [comment]
How can I obtain a newer version? Is this link below? 
https://github.com/mlpack/mlpack

**Comment by conradsnicta** [comment]
> How can I obtain a newer version? Is this link below?
> https://github.com/mlpack/mlpack

Clone the repo using git on a command line:
`$ git clone https://github.com/mlpack/mlpack.git`

**Comment by samuelbertolini** [comment]
This version is requiring cereal, and cereal gives the same error with catch2

**Comment by shrit** [comment]
@samuelbertolini use `cmake -DDOWNLOAD_DEPENDENCIES=ON ../`

**Comment by shrit** [comment]
@conradsnicta We will try to make it close before next year. The issue we can not release an immediate version because we no longer have a compatible public API. We have to go for mlpack 4.0
We only need to finish #2777 and then release a new version.

**Comment by samuelbertolini** [comment]
Thank y'all! The installation now is completed. I will close the issue.

**Comment by conradsnicta** [comment]
@shrit 
> We will try to make it close before next year. The issue we can not release an immediate release because we no longer have a compatible public API. We have to go for mlpack 4.0 We only need to finish #2777 and then release a new version.

I don't mean taking the current git master and doing a release. Instead, I mean taking mlpack-3.4.2.tar.gz (from the mlpack website), updating catch.hpp within that, disabling the tests, and releasing it as mlpack-3.4.3.

**Comment by shrit** [comment]
oh, Indeed that is a good idea, I did not think about that :+1:

**Comment by samuelbertolini** [comment]
I would also like to suggest changing the "Configuring Cmake" in the Documentation, adding the command "cmake -DDOWNLOAD_DEPENDENCIES=ON ../" . This helped a lot to install.

**Comment by samuelbertolini** [comment]
Another extra problem, but now compiling.
Sorry to come back here again, but mlpack is not compiling and not even link mlpack is soving. I am trying the following:

`g++ -g -fopenmp -L/usr/local/lib/ -lmlpack test.cpp -o test`

But the result is:

`...
/usr/local/include/mlpack/core/data/load_impl.hpp:78: undefined reference to ``mlpack::util::PrefixedOutStream::operator<<(std::ostream& (*)(std::ostream&))'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
collect2: error: ld returned 1 exit status
Build finished with error(s).`

Any idea how to solve this?

**Comment by rcurtin** [comment]
@samuelbertolini do you have multiple versions of mlpack installed on your system?  This could happen if the version of `libmlpack.so` found by the linker differs from the version of mlpack's headers that are being used.  Situations like this can arise if, e.g., `apt-get install libmlpack-dev` is done, but then *also* mlpack is compiled from source and installed to `/usr/local/`.  So, I'd suggest checking to see if there are multiple versions on your system.

**Comment by samuelbertolini** [comment]
@rcurtin , the problem was how to compile. What I did was: I copied this [Makefile](https://github.com/mlpack/examples/blob/c1e3eeb9c9997925e9598a1803e7c730ebd4d83f/mnist_cnn/Makefile), modified, and solved the problem. In any case, I checked this "multiple versions problems" from another issues, and it did not work. Now it is working. It takes a while to complile, but works. Thanks @rcurtin !
